use-cases/xtclim/hpo.py

```diff
@@ -0,0 +1,259 @@
+import argparse
+import os
+import sys
+from pathlib import Path
+from typing import Dict
+import yaml
+
+import matplotlib.pyplot as plt
+import ray
+import torch
+from ray import train, tune
+
+from itwinai.parser import ConfigParser
+
+sys.path.append(os.path.join(os.path.dirname(__file__), 'src'))
+sys.path.append(os.path.join(os.path.dirname(__file__), 'preprocessing'))
+
+
+def run_trial(config: Dict, data: Dict):
+    """Execute a single trial using the given configuration (config).
+    This runs a full training pipeline - you can also specify a pipeline as a dictionary,
+    e.g. if you only want to run certain parts without changing your config.yaml file
+    (see below).
+
+    Args:
+        config (dict): A dictionary containing hyperparameters, such as:
+            - 'batch_size' (int): The size of the batch for training.
+            - 'lr' (float): The learning rate for the optimizer.
+        data (dict): A dictionary containing a "pipeline_path" field, which points to the yaml
+            file containing the pipeline definition
+    """
+    with open('pipeline.yaml', 'r') as f:
+        yaml_config = yaml.safe_load(f)
+
+    # Override keys of hyperparameters to be tuned from Ray config
+    yaml_config['batch_size'] = config['batch_size']
+    yaml_config['lr'] = config['lr']
+
+    # Set loggers to None as Ray logs the runs
+    yaml_config['pipeline']['init_args']['steps']['training-step']['init_args']['logger'] = None
+    yaml_config['pipeline']['init_args']['steps']['evaluation-step']['init_args']['logger'] = None
+
+    seasons_list = yaml_config['seasons']
+    for season in seasons_list:
+        yaml_config['pipeline']['init_args']['steps']['training-step']['init_args']['seasons'] = season
+        model_uri = f"outputs/cvae_model_{season}1d_1memb.pth"
+        yaml_config['pipeline']['init_args']['steps']['evaluation-step']['init_args']['model_uri'] = model_uri
+        yaml_config['pipeline']['init_args']['steps']['evaluation-step']['init_args']['seasons'] = season
```

Comment:
there is an easier way of doing this, which has two steps:

First, in `pipeline.yaml` declare `season` and `model_uri` as "global variables" at the top of the config file. Then reference them in all the fields where they are needed by leveraging omegaconf variable interpolation.

```yaml
...
season: "unk"
model_uri: "unk"
...

pipeline:
   class_path: itwinai.pipeline.Pipeline
   init_args:
     steps:
         preprocessing-step:
           class_path: preprocessing.preprocess_functions_2d_ssp.PreprocessData
           init_args:
               dataset_root: ${dataset_root}
               scenario: ${scenario}
         preprocessing-split-step:
           class_path: preprocessing.preprocess_2d_seasons.SplitPreprocessedData
           init_args:
               scenario: ${scenario}
         training-step:
           class_path: src.trainer.XTClimTrainer
           init_args:
               epochs: ${epochs}
               batch_size: ${batch_size}
               lr: ${lr}
               strategy: ${strategy}
               season: ${season}
               logger:
                 class_path: itwinai.loggers.LoggersCollection
                 init_args:
                   loggers:
                     - class_path: itwinai.loggers.MLFlowLogger
                       init_args:
                         experiment_name: XTClim (Cerfacs)
                         log_freq: epoch
                     - class_path: itwinai.loggers.WandBLogger
                       init_args:
                         log_freq: epoch
         evaluation-step:
           class_path: src.anomaly.XTClimPredictor
           init_args:
             evaluation: ${evaluation}
             batch_size: ${batch_size}
             strategy: ${strategy}
             season: ${season}
             model_uri: ${model_uri}
             logger:
                 class_path: itwinai.loggers.LoggersCollection
                 init_args:
                     loggers:
                         - class_path: itwinai.loggers.MLFlowLogger
                           init_args:
                               experiment_name: XTClim (Cerfacs)
                               log_freq: epoch

```


Second, override them in python using the `override_keys` in the ConfigParser: https://itwinai.readthedocs.io/latest/python-api/itwinai.parser.html#itwinai.parser.ConfigParser